File: sdk/java/src/main/java/org/opena2a/aim/atx/CrlCache.java
package org.opena2a.aim.atx;

import java.lang.System.Logger;
import java.lang.System.Logger.Level;
import java.time.Duration;
import java.util.Objects;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.function.LongSupplier;

/**
 * Asynchronously-refreshed revocation-list (CRL) cache for local ATX verification.
 *
 * <p>Revocation is the one credential check designed to ride on a short-lived, cached
 * list (AAP §6): the decision path never makes a network call. A background timer
 * refreshes the list on an interval; {@link #current()} returns the cached list while
 * it is within its TTL and {@code null} once it goes stale.
 *
 * <p>By default a stale cache <em>soft-fails open on revocation only</em> — signature,
 * expiry and issuer-trust stay hard (see {@link RefreshingAtxVerifier}) — so a transient
 * CRL-feed outage degrades revocation freshness instead of taking down all verification.
 * The TTL bounds the window in which a revoked credential could slip through; choose
 * {@link StalePolicy#REJECT} for a fail-closed deployment.
 *
 * <p>Thread-safe: {@link #current()}/{@link #isFresh()} are lock-free reads of volatile
 * state and never throw, so they are safe on the verification decision path.
 */
public final class CrlCache {

    /** What a verifier should do when the cached CRL is stale (beyond TTL). */
    public enum StalePolicy {
        /** Skip revocation (other checks stay hard). */
        SOFT_OPEN,
        /** Fail the credential closed. */
        REJECT
    }

    /** Supplies the current CRL. Invoked only off the decision path (background timer + {@link #refreshNow()}). */
    @FunctionalInterface
    public interface CrlSource {
        AtxTrustAnchors.Crl fetch() throws Exception;
    }

    private static final Logger LOG = System.getLogger(CrlCache.class.getName());

    private final CrlSource source;
    private final long ttlMillis;
    private final long refreshIntervalMillis;
    private final StalePolicy onStale;
    private final LongSupplier nowMillis;
    private final Consumer<Throwable> onError;

    private volatile AtxTrustAnchors.Crl data;
    private volatile long fetchedAtMillis = -1L;
    private volatile Throwable lastError;
    private ScheduledExecutorService scheduler;

    private CrlCache(Builder b) {
        this.source = b.source;
        if (b.ttlMillis <= 0) {
            throw new IllegalArgumentException("ttl must be > 0");
        }
        this.ttlMillis = b.ttlMillis;
        this.refreshIntervalMillis =
                b.refreshIntervalMillis != null ? b.refreshIntervalMillis : Math.max(1L, b.ttlMillis / 2L);
        if (this.refreshIntervalMillis <= 0) {
            throw new IllegalArgumentException("refreshInterval must be > 0");
        }
        this.onStale = b.onStale;
        this.nowMillis = b.nowMillis;
        this.onError = b.onError != null
                ? b.onError
                : (t -> LOG.log(Level.WARNING, "CRL refresh failed; serving cached list until TTL", t));
    }

    /** Start building a cache around {@code source}. */
    public static Builder builder(CrlSource source) {
        return new Builder(source);
    }

    /**
     * Fetch the CRL once and update the cache. Never throws — a fetch failure is captured
     * ({@link #status()}) and the previous list is retained until it goes stale. Synchronized
     * so a manual call and the background timer cannot interleave a half-written update.
     */
    public synchronized void refreshNow() {
        try {
            AtxTrustAnchors.Crl next = source.fetch();
            if (next == null) {
                throw new IllegalStateException("CRL source returned null");
            }
            this.data = next;
            this.fetchedAtMillis = nowMillis.getAsLong();
            this.lastError = null;
        } catch (Throwable t) {
            this.lastError = t;
            onError.accept(t);
        }
    }

    /** Start the background refresh timer (idempotent) and run an immediate refresh on the timer thread. */
    public synchronized void start() {
        if (scheduler != null) {
            return;
        }
        scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "atx-crl-refresh");
            t.setDaemon(true);
            return t;
        });
        scheduler.scheduleAtFixedRate(this::refreshNow, 0L, refreshIntervalMillis, TimeUnit.MILLISECONDS);
    }

    /** Stop the background refresh timer (idempotent). */
    public synchronized void stop() {
        if (scheduler != null) {
            scheduler.shutdownNow();
            scheduler = null;
        }
    }

    /** Whether the cache holds a list that is within its TTL. */
    public boolean isFresh() {
        long at = fetchedAtMillis;
        return at >= 0 && (nowMillis.getAsLong() - at) <= ttlMillis;
    }

    /**
     * The cached CRL while fresh (within TTL), otherwise {@code null}. Lock-free and never
     * throws — safe on the decision path. A {@code null} return means the caller must apply
     * its stale policy ({@link StalePolicy}).
     */
    public AtxTrustAnchors.Crl current() {
        return isFresh() ? data : null;
    }

    /** The configured stale policy. */
    public StalePolicy onStale() {
        return onStale;
    }

    /** Observability snapshot; never throws. */
    public Status status() {
        long at = fetchedAtMillis;
        Long ageMs = at < 0 ? null : (nowMillis.getAsLong() - at);
        return new Status(data != null, ageMs, isFresh(), lastError);
    }

    /** Cache state snapshot. {@code ageMillis} is null when nothing has been loaded yet. */
    public record Status(boolean loaded, Long ageMillis, boolean fresh, Throwable lastError) {}

    /** Fluent builder for {@link CrlCache}. */
    public static final class Builder {
        private final CrlSource source;
        private long ttlMillis = 300_000L;
        private Long refreshIntervalMillis;
        private StalePolicy onStale = StalePolicy.SOFT_OPEN;
        private LongSupplier nowMillis = System::currentTimeMillis;
        private Consumer<Throwable> onError;

        private Builder(CrlSource source) {
            this.source = Objects.requireNonNull(source, "source");
        }

        /** Max age a cached CRL is served (default 5 minutes). */
        public Builder ttl(Duration ttl) {
            this.ttlMillis = Objects.requireNonNull(ttl, "ttl").toMillis();
            return this;
        }

        /** Background refresh interval (default {@code ttl/2}). */
        public Builder refreshInterval(Duration interval) {
            this.refreshIntervalMillis = Objects.requireNonNull(interval, "interval").toMillis();
            return this;
        }

        /** Stale-cache behavior (default {@link StalePolicy#SOFT_OPEN}). */
        public Builder onStale(StalePolicy policy) {
            this.onStale = Objects.requireNonNull(policy, "policy");
            return this;
        }

        /** Injectable millisecond clock (tests). Defaults to {@code System::currentTimeMillis}. */
        public Builder nowMillis(LongSupplier nowMillis) {
            this.nowMillis = Objects.requireNonNull(nowMillis, "nowMillis");
            return this;
        }

        /** Sink for refresh errors (default logs at WARNING). */
        public Builder onError(Consumer<Throwable> onError) {
            this.onError = onError;
            return this;
        }

        public CrlCache build() {
            return new CrlCache(this);
        }
    }
}

File: sdk/typescript/src/local/CrlCache.ts
/**
 * Asynchronously-refreshed revocation-list (CRL) cache for local ATX verification.
 *
 * Revocation is the one credential check designed to ride on a short-lived, cached
 * list (AAP §6): the decision path never makes a network call. A background timer
 * refreshes the list on an interval; {@link CrlCache.current} returns the cached
 * list while it is within its TTL and `null` once it goes stale.
 *
 * By default a stale cache *soft-fails open on revocation only* — signature, expiry
 * and issuer-trust stay hard — so a transient CRL-feed outage degrades revocation
 * freshness instead of taking down all verification. The TTL bounds the window in
 * which a revoked credential could slip through; set `onStale: 'reject'` for a
 * fail-closed deployment where availability-over-revocation-freshness is not
 * acceptable.
 */

/** A federated certificate revocation list. */
export interface CrlData {
  entries: Array<{ agentId: string; reason?: string }>;
}

/** What a verifier should do when the cached CRL is stale (beyond TTL). */
export type CrlStalePolicy = 'soft-open' | 'reject';

export interface CrlCacheConfig {
  /** Fetches the current CRL. Invoked only off the decision path (background timer + `refreshNow`). */
  fetch: () => Promise<CrlData>;
  /** Max age, in ms, that a cached CRL is served. After this, `current()` returns null. Default 300_000 (5 min). */
  ttlMs?: number;
  /** Background refresh interval, in ms. Default `max(1, floor(ttlMs / 2))`. */
  refreshIntervalMs?: number;
  /** Behavior when the cache is stale: `'soft-open'` (default) skips revocation; `'reject'` fails closed. */
  onStale?: CrlStalePolicy;
  /** Injectable clock (tests). Defaults to `Date.now`. */
  now?: () => number;
  /** Sink for refresh errors / degraded-state notices. Defaults to `console.warn`. */
  onError?: (err: unknown) => void;
}

export interface CrlCacheStatus {
  /** A list has been successfully fetched at least once. */
  loaded: boolean;
  /** Age of the cached list in ms, or null if never loaded. */
  ageMs: number | null;
  /** Loaded and within TTL. */
  fresh: boolean;
  /** The most recent refresh error, or null. */
  lastError: unknown;
}

export class CrlCache {
  private readonly fetchFn: () => Promise<CrlData>;
  private readonly ttlMs: number;
  private readonly refreshIntervalMs: number;
  /** Policy a verifier applies when this cache is stale. */
  readonly onStale: CrlStalePolicy;
  private readonly now: () => number;
  private readonly onError: (err: unknown) => void;

  private data: CrlData | null = null;
  private fetchedAt: number | null = null;
  private lastError: unknown = null;
  private timer: ReturnType<typeof setInterval> | null = null;
  private inFlight: Promise<void> | null = null;

  constructor(config: CrlCacheConfig) {
    if (typeof config.fetch !== 'function') {
      throw new TypeError('CrlCache requires a fetch function');
    }
    this.fetchFn = config.fetch;
    this.ttlMs = config.ttlMs ?? 300_000;
    if (!(this.ttlMs > 0)) {
      throw new RangeError('CrlCache ttlMs must be > 0');
    }
    this.refreshIntervalMs = config.refreshIntervalMs ?? Math.max(1, Math.floor(this.ttlMs / 2));
    if (!(this.refreshIntervalMs > 0)) {
      throw new RangeError('CrlCache refreshIntervalMs must be > 0');
    }
    this.onStale = config.onStale ?? 'soft-open';
    this.now = config.now ?? Date.now;
    this.onError =
      config.onError ??
      ((err) => console.warn('[CrlCache] refresh failed; serving cached list until TTL:', err));
  }

  /**
   * Fetch the CRL once and update the cache. Resolves even when the fetch fails — the
   * error is captured ({@link status}) and the previous list is retained until it goes
   * stale. Concurrent calls are coalesced into a single in-flight fetch.
   */
  async refreshNow(): Promise<void> {
    if (this.inFlight) {
      return this.inFlight;
    }
    this.inFlight = (async () => {
      try {
        const next = await this.fetchFn();
        if (!next || !Array.isArray(next.entries)) {
          throw new TypeError('CRL fetch returned an invalid shape (expected { entries: [...] })');
        }
        this.data = next;
        this.fetchedAt = this.now();
        this.lastError = null;
      } catch (err) {
        this.lastError = err;
        this.onError(err);
      } finally {
        this.inFlight = null;
      }
    })();
    return this.inFlight;
  }

  /**
   * Start the background refresh timer (idempotent) and kick off an immediate refresh.
   * On Node the timer is `unref`-ed so it never keeps the process alive on its own.
   */
  start(): void {
    if (this.timer) {
      return;
    }
    void this.refreshNow();
    this.timer = setInterval(() => void this.refreshNow(), this.refreshIntervalMs);
    (this.timer as { unref?: () => void } | undefined)?.unref?.();
  }

  /** Stop the background refresh timer (idempotent). */
  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  /** Whether the cache holds a list that is within its TTL. */
  isFresh(): boolean {
    return this.fetchedAt !== null && this.now() - this.fetchedAt <= this.ttlMs;
  }

  /**
   * The cached CRL while fresh (within TTL), otherwise `null`. Synchronous and never
   * throws — safe on the verification decision path. A `null` return means the caller
   * must apply its {@link onStale} policy (soft-open: skip revocation; reject: fail closed).
   */
  current(): CrlData | null {
    return this.isFresh() ? this.data : null;
  }

  /** Observability snapshot; never throws. */
  status(): CrlCacheStatus {
    const ageMs = this.fetchedAt === null ? null : this.now() - this.fetchedAt;
    return { loaded: this.data !== null, ageMs, fresh: this.isFresh(), lastError: this.lastError };
  }
}
